### PR TITLE
feat(blockchain-tree): add `Display` for blocks in `Chain`

### LINKED ISSUE #3204: Improve `blockchain_tree: Committing new canonical chain` message
### Describe the feature

the new chain can include many blocks, depending how many we buffered:

> Committing new canonical chain: [(3704247, 0xdd13ba166c3ee4f4a53c24978c43ec05f3652c56e32f902a245f772da06340db), (3704248, 0x4dcb1024bf8916d7b1fb6835976e8e63ca2e73ba26ce66e4ae5eba9da7bb09b0), (3704249, 0x81fe7d74e63fe73f7c51519095b74800113b823fcdd59c344e03c02e690d65b4), (3704250, 0xf396f8406d8b16d7b011522597bb93bc771ab087405f245d1fb8b9893af0a7e1), (3704251, 0xd5c319482b128afb20c8637ee593b0a0f48c36c8a08986121ee652a28dc7222d), (3704252, 0xa1191f09f7cb5c483c40248d3f270a8cdf5c04ff658d5d686c886de1ebb25e35), (3704253, 0x3c15666782dba8d67df6e42396891f61afe11f68bc9eed03826e0e1fc0bb58db), (3704254, 0x877d03c2377b271d424683293d6b1bc99994188b8092e3f0bbd857d258d28821), (3704255, 0xd18222c5dfc298db069aa878600d60ed521a712d268d3f4456bc9a2e47efbc4e), (3704256, 0xec4a6a792511b724a12e86d644ef082a7bde84f851708a5db08cbb1f3ad4c14d), (3704257, 0xe95b0463298cd80c433fe6697183145cf6ebea8fa4a9ea08b942f763a69c9f44), (3704258, 0xfd206d7666a41b6504139db6b48e1debb4eed947e3fac4552d98d169da8202c6), (3704259, 0xcae9e4e192899d9be623b51707215797edf9d752193e8d36477f076397644bdc), (3704260, 0xc8eb0b4a6fbe4e98c8a74194b413232f6880379499843042dc0c95870cb5fce2), (3704261, 0x3bb4646d0717ce3c6c84ed883a8c6a33d16cdb7f1b3406a5f0b2837bb66a7e49), (3704262, 0xe807955a7a2ef1cec824dec116f676325bc42737ef73b2aff3501f708388e562), (3704263, 0x80ba4291da6c69f88a959b9d64b84431c722d13843ce070bc82bc0606a234209), (3704264, 0xbc75bfaa0906e183a83a67e236bf3a72dd60afe47079d6e0027533370bf5b989), (3704265, 0x84204e2aa3cd3b4badbd1cf6ddcc8983d50f47e04331b86caca927b71dbb902a), (3704266, 0xb6ebda6f7fc591496e4030d1f9be15ea5cb0c74273d1a266b3b6294d3328f6ae), (3704267, 0x864a8444cbbc20ece8fdbae4514a0987a04b38092ec014d8fb47c030b739a7b4), (3704268, 0xff81a52e7c538969bc73c6fef6d6938d1840adacec5e3ea40278fc71dc6f6c0d), (3704269, 0xfbb7959bb0c0e141188136ad95f5a509dcb093647752bb675ac89db0b959af29), (3704270, 0xc771afdce9ad4151e9ae671bf3da9ef5b3654a6a4e71af9ee73df1247b1aafc5), (3704271, 0xeb93682d376b8938a8a2ad93506fbccd912309e6f1c75164515150cdd8bdc922), (3704272, 0xadf2bf711e5da2524cb98c0fd9b06c9d2cef31e152155b7db6b51957cc6552cd), (3704273, 0xaa5b7adf56774cd9976f7c2cbd62189768504a184ee8e351be6c2c06d5356567), (3704274, 0xe0a07d945e984824312940eacded0a4b2e1d50b6a10ca5f802f588576e2733e1), (3704275, 0xba14fec0f466348a9f491f5c0e8242ee3757e0343c44826c0d312eeb7b2c850f), (3704276, 0x714a0c86264a56863a214c0d70ba6819572f7bdcd55138713d17be38dbebb535), (3704277, 0x9f87477c626218f584b84e817556b92fba1c88286cb3b50fd301669dec9b1cd5), (3704278, 0x5a208a703539a614dd7ec7d820de195b762e4c6b31d18ccaa4fd89bc0ddcbf23), (3704279, 0xf4be8b7fef0aad90937558c60a3548a3ea7a5b72e490b2db2dda1432b2be86fd), (3704280, 0xd1e6582ed99904934bc616f4dd9922370b672895ca88cf48f5cb7c6e04ca510a), (3704281, 0x4445105a29373b18cc769bccea71fc3bf70e1525c58d5cc8f4d120a2bf906929), (3704282, 0xd23d71db5d55aaffbd1ccd473fa0bedc3446554a71c46a9666564673dba151c1), (3704283, 0x5fbbd18b1d8266d671a2426a81f06cfa9263ef1c26f2f8a245c6fb6420316108), (3704284, 0x51e9d09fdbea94bea6514bb61bbf2cac7abf45fcc959aff39a6dabb6c0d035d2), (3704285, 0x32b7a71eda34c866c42ec61f15e247f6f85c549624b90fcb4520d74b7cbb86ba), (3704286, 0x0d9730cb3c13f4c4a62399b413044ddcb3b3babd6052706c417520f9c10375b0), (3704287, 0x966a6dcaef8f8376cffb54d3e1f9d6cd550c820f340644a440a4a5bd929a0170), (3704288, 0x63684cad559040aaa5c53b831f3bb5e968235b02063501dfdfac849a76195581), (3704289, 0x6f5a5fa464d2bd40e0312f4b94f015ae9a69f8734838de2e6fccad7090a0687d), (3704290, 0x6b6a75616e074a09213433a34a80fa634d44e51850c5b642f9d980930326e696), (3704291, 0xf04e20ac7c5a7c1052cc763f1d25940f411bd3288dc8396598a7b3429caae4f1), (3704292, 0x43977eebc4bf4afe30a78de2cc7a77fd411fbb594c54fe8a5c597d8452bbe10f), (3704293, 0x46d45cf3c955f611152b2aaa92fa3bf653099bdfc05580c620573e5cb14d4952), (3704294, 0xa72544e1b17f0b5921224729c75f8579992b49265635042630e2a325f3827fe2), (3704295, 0xa681bbf7e11584141040fdf8a70cfdaa5365e3be2b09eedc45964aa6e27aa815), (3704296, 0x46161f6a45c871587600592f8896208d00d6b5adcb80b2635305bfab5383fa60), (3704297, 0xf058f11674150fd67344d3a5a8cf74b818c01692fc40ac81ec8bd003a39de9cc), (3704298, 0x947dcc56590d0c016a3c6ed824970df3bd95396f97b9f0beb07316f7ce4a887f), (3704299, 0x2ed8bf5502cb7ea6f560f5eac6e4a7d4a62e0b6a54fc58c56c51ddb039cc8b55), (3704300, 0x8ad9afe89021204f62c73e517d59882dff5b4906c46f87342f863cb951abfd3d), (3704301, 0x551cd7e74a39d299a525ca21b9d0881f669005831851025265e571cac9c2fed8), (3704302, 0x0663854544bc2ff3da7e36e5c1365cc2cc98af63916ced5bef537c64146a1abc), (3704303, 0x5a096b6ee1fd2928adbf895dcb1922ed30f488fd9ec21a35c15f94bd0f63664d), (3704304, 0x7aa236569b480681854ba239163a009e6a39a23c65c155e204b253f24bd673a2), (3704305, 0x3fa50990bdf22fbe378d235ee5b61c5dee4536a13b9ea197e7534a127f6e9cb5), (3704306, 0xe99f6a8a54028f305ab1d80e982d952b6a6ddfe668a20f73e9c7ffbf3023763d), (3704307, 0x8c3562b0835f29a263c824a28394d9c503bd8251debbd3541d42881edfaaaa7c), (3704308, 0xace2ba15732295e027f1829fc50629dbf2a50750dbf4f60bddd5d0620501178f), (3704309, 0x43b89ce4dc59997670e1fa4dca62c00adbece49afc3ac9ef2d190a83d38dac13), (3704310, 0x0d57420b4892a428278ac732e4067d957bd7faa9183fa862f6ca31bb84a9f4d4), (3704311, 0x06aee1187d822aacd4393f211cd9311e50b915375dd4381ef8457beed9dcd051), (3704312, 0xceb36a3aca64430a24182313aaac1d88aa90717c305d6efd731c13303461bce5), (3704313, 0x9499ef49487fb6f228de2d201611cebf5f8c01ba04c8d7659d140b8b8b9d8cba)]

https://github.com/paradigmxyz/reth/blob/b90d0b41069015d5746b08b3e408e738bc65a6a7/crates/blockchain-tree/src/blockchain_tree.rs#L924-L928

## TODO

I think we should add a helper wrapper type with a new Display impl that prints this information and truncates blocks if the chain is larger than x blocks (3?)

it should include
`[first block, (second block or truncated ...) , last block]`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tcoratger**
Should solve #3204.

### CODE REVIEW

**Comment by onbjerg** [comment]
What does this look like in practice?

File: crates/storage/provider/src/chain.rs [hunk 1]
```diff
@@ -217,6 +217,37 @@ impl Chain {
...
 
+/// Wrapper type for `blocks` display in `Chain`
+pub struct DisplayBlocksChain<'a>(pub &'a BTreeMap<BlockNumber, SealedBlockWithSenders>);
+
+impl<'a> fmt::Display for DisplayBlocksChain<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let block_len = self.0.len();
+
+        if block_len == 0 {
+            return Ok(())
```
Comment (mattsse) [inline crates/storage/provider/src/chain.rs:228]: should write like an empty vec: `[]`

File: crates/storage/provider/src/chain.rs [hunk 2]
```diff
@@ -217,6 +217,37 @@ impl Chain {
...
+                let block_num_hash = block.1.num_hash();
+                write!(
+                    f,
+                    "({:?}, {:?}...), ",
+                    block_num_hash.number,
+                    block_num_hash.hash.to_string().truncate(10)
+                )?;
+            }
+            write!(f, "{:?}", self.0.iter().last().unwrap().1.num_hash())?;
+        }
```
Comment (mattsse) [inline crates/storage/provider/src/chain.rs:244]: what I meant by truncate was truncating all additional entries if len > 3:

`[(first),... (last)]`

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, smol nit and would like a sanity test that uses `to_string` and compares against the expected output.

File: crates/storage/provider/src/chain.rs (hunk 1 above)
Comment (tcoratger) [reply crates/storage/provider/src/chain.rs]: @mattsse Done

File: crates/storage/provider/src/chain.rs (hunk 2 above)
Comment (tcoratger) [reply crates/storage/provider/src/chain.rs]: @mattsse Sorry I misunderstood, I thought you wanted to display only a few characters of the hash in this case. I just modified.

**Review by onbjerg — APPROVED** [review]
lgtm, pending @mattsse

File: crates/storage/provider/src/chain.rs [hunk 3]
```diff
@@ -217,6 +217,34 @@ impl Chain {
...
+        if self.0.len() <= 3 {
+            write!(f, "[")?;
+            let mut iter = self.0.values().map(|block| block.num_hash());
+            if let Some(block_num_hash) = iter.next() {
+                write!(f, "{:?}", block_num_hash)?;
+                for block_num_hash_iter in iter {
+                    write!(f, ", {:?}", block_num_hash_iter)?;
+                }
+            }
+            write!(f, "]")?;
```
Comment (mattsse) [inline crates/storage/provider/src/chain.rs:234]: nice!

**Review by mattsse — APPROVED** [review]
awesome!

**Pull request merged**